File: packages/domain/src/milpbooklm_domain/notebook.py
"""Notebook entity and its status state machine (pure, stdlib-only)."""

from __future__ import annotations

import uuid
from dataclasses import dataclass
from enum import StrEnum


class NotebookStatus(StrEnum):
    """Lifecycle states of a notebook."""

    ACTIVE = "active"
    ARCHIVED = "archived"


@dataclass(frozen=True, slots=True)
class Notebook:
    """A research notebook aggregate root (immutable value object)."""

    id: uuid.UUID
    title: str
    status: NotebookStatus = NotebookStatus.ACTIVE

    def archive(self) -> Notebook:
        """Return a new archived notebook; the original is unchanged."""
        if self.status is NotebookStatus.ARCHIVED:
            return self
        return Notebook(id=self.id, title=self.title, status=NotebookStatus.ARCHIVED)

File: packages/web/domain/src/notebook.ts
/**
 * Frontend domain: pure types and state machines.
 *
 * This package imports nothing beyond itself. It mirrors the Python
 * `milpbooklm_domain` entity so the web app never drifts from the backend
 * model (ARCH-01-001 scope: the domain is the shared boundary).
 */

/** Lifecycle states of a notebook. */
export const NotebookStatus = {
  Active: "active",
  Archived: "archived",
} as const;

export type NotebookStatus = (typeof NotebookStatus)[keyof typeof NotebookStatus];

/** A research notebook (immutable value object). */
export interface Notebook {
  readonly id: string;
  readonly title: string;
  readonly status: NotebookStatus;
}

/** Pure state transition: returns a new archived notebook (idempotent). */
export function archive(notebook: Notebook): Notebook {
  if (notebook.status === NotebookStatus.Archived) {
    return notebook;
  }
  return { ...notebook, status: NotebookStatus.Archived };
}
